### PR TITLE
fix: [CDS-81532]: Handle service now response in json format

### PR DESCRIPTION
**Pull request by vivekkarnatiharness**
## Describe your changes
Handle servicenow exception and throw proper exception instead of json format.
image
image
image

## Checklist
- [x] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/55255)

### CODE REVIEW

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowErrorResponse.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    JsonNode jsonNode2 = node.get("error");
+    ServiceNowError serviceNowErrorResponse = new ServiceNowError(jsonNode2);
+    String parsedErrorDetails = serviceNowErrorResponse.getErrorMessage();
+    if (!StringUtils.isBlank(status)) {
+      this.status = status;
+      this.errorMessage = parsedErrorDetails;
+      return;
+    }
+
+    throwRefreshTokenException("error response doesn't have \"error\" or \"errorCode\" field");
```
Comment (naman-goenka) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowErrorResponse.java:38]: We need to throw snow exception here

Comment (vivekkarnatiharness) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowErrorResponse.java]: Done, changed it

**Comment by vivekkarnatiharness** [comment]
trigger uts

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowErrorResponse.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+@JsonIgnoreProperties(ignoreUnknown = true)
+@Slf4j
+public class ServiceNowErrorResponse {
+  String status;
+  String errorMessage;
+
+  public ServiceNowErrorResponse(JsonNode node) throws JsonProcessingException {
+    String status = JsonNodeUtils.getString(node, "status");
+    JsonNode jsonNode2 = node.get("error");
+    ServiceNowError serviceNowErrorResponse = new ServiceNowError(jsonNode2);
```
Comment (naman-goenka) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowErrorResponse.java:32]: a null check here would be required

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowErrorResponse.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    String parsedErrorDetails = serviceNowErrorResponse.getErrorMessage();
+    if (!StringUtils.isBlank(status)) {
+      this.status = status;
+      this.errorMessage = parsedErrorDetails;
+      return;
+    }
+
+    throw new ServiceNowException("error response doesn't have \"error\" or \"errorCode\" field",
```
Comment (naman-goenka) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowErrorResponse.java:40]: we are parsing for status and error field

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowError.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+@OwnedBy(CDC)
+@Data
+@FieldDefaults(level = AccessLevel.PRIVATE)
+@JsonIgnoreProperties(ignoreUnknown = true)
+@Slf4j
+public class ServiceNowError {
+  String errorMessage;
+  public ServiceNowError(JsonNode node) {
+    String message = JsonNodeUtils.getString(node, "message");
+    String detail = JsonNodeUtils.getString(node, "detail");
```
Comment (naman-goenka) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowError.java:24]: if both are blank then we can throw an exception and hence default to raw exception

**Review by naman-goenka — COMMENTED** [review]
Can we please attach some screenshots of the improved error messages in acl missing or bad request snow API calls

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowError.java (hunk 4 above)
Comment (vivekkarnatiharness) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowError.java:24]: Done

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowErrorResponse.java (hunk 3 above)
Comment (vivekkarnatiharness) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowErrorResponse.java:40]: Done

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowErrorResponse.java (hunk 2 above)
Comment (vivekkarnatiharness) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowErrorResponse.java:32]: Done

**Review by naman-goenka — APPROVED** [review]

**Comment by vivekkarnatiharness** [comment]
trigger uts

**Review by abhinavmittal-harness — APPROVED** [review]
> Reviewed 2 of 4 files at r1, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55255)* status: 4 file(s) to review. 4. unresolved discussions.

**Comment by vivekkarnatiharness** [comment]
trigger review

**Comment by vivekkarnatiharness** [comment]
trigger messagecheck

**Pull request merged**